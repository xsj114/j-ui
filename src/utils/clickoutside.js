import {on} from '@utils/utils'

let nodeList = []

let startClick

let seed = 0

const ctx = '@@clickoutside'


on(document,'mousedown',e => (startClick = e) )
on(document,'mouseup',e => {
    nodeList.forEach(node => node[ctx].documentHandle(e, startClick))
})

export default {
    bind (el, binding, vnode)  {
        nodeList.push(el)
        const id = seed++
        el[ctx] = {
            id,
            documentHandle: createDocumentHandle(el, binding, vnode),
            bindingFn: binding.value,
            methodName: binding.expression
        }
    },
    unbind (el) {
        let len = nodeList.length
        for (let i = 0; i < len; i++) {
            if (nodeList[i][ctx].id === el[ctx].id) {
                nodeList.splice(i, 1)
                break
            }
        }
        delete el[ctx]
    }
}

const  createDocumentHandle =  (el, binding, vnode) => {
    return function (mouseup, mousedown) {
        if (
            !vnode ||
            !vnode.context||
            !mouseup.target ||
            !mousedown.target ||
            el.contains(mouseup.target) ||
            el.contains(mousedown.target) ||
            (vnode.context.popperElement && 
            (vnode.context.popperElement.contains(mousedown.target) || vnode.context.popperElement.contains(mouseup.target))
            )
        ) { return }

        if (binding.expression && el[ctx].methodName && vnode.context[el[ctx].methodName]) {
            // 测试时使用
            vnode.context[el[ctx].methodName]()
        } else {
            el[ctx].bindingFn && el[ctx].bindingFn()
        }
    }
}



