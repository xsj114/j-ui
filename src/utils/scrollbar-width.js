export default () => {

    let scrollBarWidth

    const outer = document.createElement('div')
    outer.style.visibility = 'hidden'
    outer.style.width = '100px'
    outer.style.position = 'absolute'
    outer.style.top = '-9999px'
    document.body.appendChild(outer)

    let widthNoScroll = outer.offsetWidth
    outer.style.overflow = 'scroll'

    let inner = document.createElement('div')
    inner.style.width = '100%'
    outer.appendChild(inner)

    let widthWithScroll = inner.offsetWidth
    outer.parentNode.removeChild(outer)

    scrollBarWidth = widthNoScroll - widthWithScroll

    return scrollBarWidth

}
