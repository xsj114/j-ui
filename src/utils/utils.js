// 获取指定对象指定属性的值
export const getValueByPath = function (object, prop) {
    prop = prop || ''
    const paths = prop.split('.')
    let current = object
    let result = null
    for (let i = 0; i < paths.length; i++) {
        const path = paths[i]
        if (!current)  break
        if (i === paths.length - 1 ) {
            result = current[path]
            break
        }
        current = current[path]
    }
    return result
}


// 判断是不是对象
export const isObject = function (value) {
    return Object.prototype.toString.call(value).toLowerCase() === '[object object]'
}

// 判断是不是null
export const isNull = function (value) {
    return Object.prototype.toString.call(value).toLowerCase() === '[object null]'
}

// 判断是不是undefined
export const isUndefined = function (value) {
    return Object.prototype.toString.call(value).toLowerCase() === '[object undefined]'
}


// 随机生成id
export const generateId = () => {
    return Math.floor(Math.random() * 100000)
}

// 防抖函数 
export const debounce = function (delay, callback) { 
    let timeoutId = null
    function wrapper() {
        var self = this
        let args = arguments
        function exec () {
            callback.apply(self,args)
        }
        if (timeoutId) {
            clearTimeout(timeoutId)
        }
        timeoutId = setTimeout(exec,delay)
    }
    return wrapper
}

// 事件监听
export const on = function (element, event, handler) {
    if (document.addEventListener) {
        if (element && event && handler) {
            element.addEventListener(event,handler,false)
        }
    } else {
        if (element && event && handler) {
            element.attachEvent('on' + event, handler)
        }
    }
}

// 取消事件监听
export const off = function (element, event, handler) {
    if (document.removeEventListener) {
        if (element && event && handler) {
            element.removeEventListener(event, handler, false)
        }
    } else {
        if (element && event) {
            element.detachEvent('on' + event, handler)
        }    
    }
}


// 两个数组内容是否相等
export const valueEquals = (a, b) => {
    if (a === b ) return true
    if ( !(a instanceof Array) ) return false
    if ( !(b instanceof Array) ) return false
    if (a.length !== b.length) return false
    for (let i = 0; i < a.length ; i++) {
        if (a[i] !== b[i]) return false
    }
    return true
}
