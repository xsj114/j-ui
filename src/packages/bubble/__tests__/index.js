import { mount } from '@vue/test-utils'
import JBubble from '../index'
import {debounce} from '@utils/utils'

describe('JBubble', () => {

    describe('create', () => {

        test('create',() => {

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })
        
            const wrapperElement = wrapper.find('.j_bubble')
            expect(wrapperElement.exists()).toBe(true)

            const firstElement = wrapper.find('span')
            expect(firstElement.html()).toBe('<span>测试</span>')

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.classes('j_bubble_popper')).toBe(true)
            expect(popperElement.classes('j_popper')).toBe(true)
            expect(popperElement.classes('j_popper_dark')).toBe(true)
            expect(popperElement.isVisible()).toBe(false)

            const ulElement = wrapper.find('ul')
            expect(ulElement.html()).toBe('<ul></ul>')

            expect(wrapper).toMatchSnapshot()
        })

        test('create and no root element', () => {

            const wrapper =  mount(JBubble)
            expect(wrapper.html()).toBe('')

            expect(wrapper).toMatchSnapshot()
            
        })


        test('create and no tag element', () => {

            const WrapperComp = {
                template: `
                <JBubble>
                    测试
                </JBubble>
              `,
                components: {
                    JBubble
                }
            }
            const wrapper = mount(WrapperComp).findComponent(JBubble)
            expect(wrapper.html()).toBe('')

            expect(wrapper).toMatchSnapshot()
        })


        test('reference element mouseover', async () => {

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })
            const referenceElement = wrapper.find('span')
            await referenceElement.trigger('mouseover')
            expect(wrapper.vm.showPopper).toBe(true)
            expect(wrapper.vm.expectedState).toBe(true)

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.isVisible()).toBe(true)

            expect(wrapper).toMatchSnapshot()
        })

        test('reference element mouseout', async () => {
            
            const fn = jest.fn()
            jest.useFakeTimers()

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })
            const referenceElement = wrapper.find('span')
            await referenceElement.trigger('mouseout')
            expect(wrapper.vm.expectedState).toBe(false)

            debounce(200, fn)
            jest.runAllTimers()

            expect(wrapper.vm.showPopper).toBe(false)

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.isVisible()).toBe(false)

            expect(wrapper).toMatchSnapshot()
        })

        test('popper element mouseover', async () => {

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })

            const popperElement = wrapper.find('.j_bubble_popper')
            await popperElement.trigger('mouseenter')
            expect(wrapper.vm.showPopper).toBe(true)
            expect(wrapper.vm.expectedState).toBe(true)

            expect(wrapper).toMatchSnapshot()
        })

        test('popper element mouseout', async () => {

            const fn = jest.fn()
            jest.useFakeTimers()

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })
            const popperElement = wrapper.find('.j_bubble_popper')
            await popperElement.trigger('mouseleave')
            expect(wrapper.vm.expectedState).toBe(false)

            debounce(200, fn)
            jest.runAllTimers()

            expect(wrapper.vm.showPopper).toBe(false)
            expect(popperElement.isVisible()).toBe(false)

            expect(wrapper).toMatchSnapshot()
        })


        test('destroyed', () => {

            const wrapper =  mount(JBubble, {
                attachTo: document.body,
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })

            expect(document.body).not.toBeEmptyDOMElement()

            expect(wrapper).toMatchSnapshot()
            wrapper.destroy()

            expect(document.body).not.toContainElement(wrapper.element)

        })

    })




    describe('size', () => {

        test('small size', () => {

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                },
                propsData: {
                    size: 'small'
                }
            })

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.classes('j_bubble_popper_small')).toBe(true)

            expect(wrapper).toMatchSnapshot()
        })


        test('large size', () => {

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                },
                propsData: {
                    size: 'large'
                }
            })

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.classes('j_bubble_popper_large')).toBe(true)

            expect(wrapper).toMatchSnapshot()
        })

    })


    describe('slot content', () => {

        test('content', async () => {

            const WrapperComp = {
                template: `
                <JBubble>
                    <span>测试</span>
                    <template v-slot:content>
                        <li>阳性</li><li>阴性</li>
                    </template>
                </JBubble>
              `,
                components: {
                    JBubble
                }
            }

            const wrapper = mount(WrapperComp).findComponent(JBubble)

            const ulElement = wrapper.find('ul')
            expect(ulElement.element).toContainHTML(`<li>阳性</li><li>阴性</li>`)

            expect(wrapper).toMatchSnapshot()
        })
    })


    describe('JBubble method', () => {

        test('open', async () => {


            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })

            wrapper.vm.open()
            expect(wrapper.vm.showPopper).toBe(true)

            await wrapper.vm.$nextTick()

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.isVisible()).toBe(true)

            expect(wrapper).toMatchSnapshot()

        })

        test('close', async () => {

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })

            wrapper.vm.close()
            expect(wrapper.vm.showPopper).toBe(false)

            await wrapper.vm.$nextTick()

            const popperElement = wrapper.find('.j_bubble_popper')
            expect(popperElement.isVisible()).toBe(false)

            expect(wrapper).toMatchSnapshot()
        })

    })


    describe('user operation', () => {

        test('move the reference element into the popper element at the moment', async () => {

            const fn = jest.fn()
            jest.useFakeTimers()

            const wrapper =  mount(JBubble, {
                scopedSlots: {
                    default: '<span>测试</span>'
                }
            })
            const referenceElement = wrapper.find('span')
            const popperElement = wrapper.find('.j_bubble_popper')
            
            await referenceElement.trigger('mouseover')
            expect(popperElement.isVisible()).toBe(true)
            await referenceElement.trigger('mouseout')
            expect(popperElement.isVisible()).toBe(true)
            debounce(100,fn)
            await popperElement.trigger('mouseenter')
            expect(popperElement.isVisible()).toBe(true)
            jest.runAllTimers()
            await popperElement.trigger('mouseleave')
            debounce(100,fn)
            await referenceElement.trigger('mouseover')
            jest.runAllTimers()
            expect(popperElement.isVisible()).toBe(true)
            await referenceElement.trigger('mouseout')
            debounce(300,fn)
            jest.runAllTimers()
            await wrapper.vm.$nextTick()
            expect(popperElement.isVisible()).toBe(false)

            expect(wrapper).toMatchSnapshot()
        }) 

    })

})
