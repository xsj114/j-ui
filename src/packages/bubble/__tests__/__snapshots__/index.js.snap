// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JBubble JBubble method close 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble JBubble method open 1`] = `<div class="j_bubble"><span>测试</span></div>`;

exports[`JBubble create create 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble create create and no root element 1`] = ``;

exports[`JBubble create create and no tag element 1`] = ``;

exports[`JBubble create destroyed 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble create popper element mouseout 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble create popper element mouseover 1`] = `<div class="j_bubble"><span>测试</span></div>`;

exports[`JBubble create reference element mouseout 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble create reference element mouseover 1`] = `<div class="j_bubble"><span>测试</span></div>`;

exports[`JBubble size large size 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark j_bubble_popper_large" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble size small size 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark j_bubble_popper_small" style="display: none;">
    <ul></ul>
  </div>
</div>
`;

exports[`JBubble slot content content 1`] = `
<div class="j_bubble"><span>测试</span>
  <div class="j_bubble_popper j_popper j_popper_dark" style="display: none;">
    <ul>
      <li>阳性</li>
      <li>阴性</li>
    </ul>
  </div>
</div>
`;

exports[`JBubble user operation move the reference element into the popper element at the moment 1`] = `<div class="j_bubble"><span>测试</span></div>`;
